[Feature] 관리자 생성

sangxxjin opened the pull request:
## 🌿 반영 브랜치
security/admin -> main

## 🛠 작업 내용
- BaseEntity의 변수명 수정 및 updatedAt의 어노테이션 변경
- 관리자 테이블 설정
- user domain을 admin으로 변경
- 관리자 계정 생성 로직 구현
- DevInitData에 관리자 생성 로직 추가
- 간단한 테스트 코드 구현

## 🔗 관련 이슈
Closes #1 

## 📋참고사항
 .env 파일에 아래와 같이 데이터 추가해야 합니다.
```
ADMIN_USERNAME= 
ADMIN_PASSWORD=
```

## 💬 리뷰 요구사항 (선택)
- admin에 getter를 지양하고 싶은데 테스트 코드를 작성하다보니 getter를 사용하였습니다. 지양하려면 어떤 방식으로 해야할까요?
- '네이밍이 아쉽거나나 테스트코드가 이게 더 있었으면 좋겠다.' 등 의견 있으시면 자유롭게 남겨주세요.

- 의견 많이 남겨주셔서 소통해주시면 좋습니다. ex) 이건 왜 이런식으로 구현했나요?

sangxxjin committed:
feat: Admin entity
본문: admin 테이블 구현

sangxxjin committed:
fix: modifiedDate의 어노테이션 수정
본문: @CreatedDate -> @LastModifiedDate으로 변경하여 정보가 업데이트 될때 데이터 업데이트 되게 수정했습니다.

sangxxjin committed:
refactor: BaseEntity 상속

sangxxjin committed:
feat: AdminRepository

sangxxjin committed:
refactor: Admin Builder 사용
본문: Getter, Setter 지양하기 위해 Builder 사용

sangxxjin committed:
feat: 비밀번호 암호화 Bean 등록

sangxxjin committed:
feat: 관리자 계정 생성 로직 구현

sangxxjin committed:
rename: user->admin

sangxxjin committed:
feat: existsByUsername

sangxxjin committed:
feat: createAdminAccountIfNotExist

sangxxjin committed:
feat: 프로그램 시작 시 admin 계정 생성

sangxxjin committed:
refactor: baseEntity 변수명 변경
본문:
createdDate -> createdAt
modifiedDate -> updatedAt

sangxxjin committed:
test: adminService의 createAdmin테스트
본문:
데이터 삽입이 정상적으로 이루어지는지 테스트

sangxxjin committed:
test: 저장된 admin의 데이터 검증

sangxxjin committed:
refactor: admin 계정 정보 env로 이동

sangxxjin committed:
refactor: toString 제거
본문:
테스트코드에서 사용하려고 작성했다가 getter를 사용하였음으로 제거했습니다.

sangxxjin committed:
refactor: 테이블명 복수형으로 변경

ChoiHyunSan commented on backend/src/main/java/com/ll/cafeservice/entity/base/BaseEntity.java:
```diff
@@ -13,8 +14,8 @@
 
-    @CreatedDate
-    private LocalDateTime modifiedDate;
+    @LastModifiedDate
+    private LocalDateTime updatedAt;
```
제가 놓쳤던 부분인데도 잘 찾아서 수정해주셔서 감사합니다 ㅠ

ChoiHyunSan commented on backend/src/main/java/com/ll/cafeservice/entity/admin/Admin.java:
```diff
@@ -0,0 +1,38 @@
+
+@Getter
+@Builder
+@NoArgsConstructor
+@AllArgsConstructor
```
이 부분은 빌더 패턴을 이용해서 생성하는 정책을 사용중이시니까  
@NoArgsConstructor(access = AccessLevel.PROTECTED)
@AllArgsConstructor(access = AccessLevel.PROTECTED)
위처럼 접근 제어자를 세부적으로 설정해주시면 좀 더 단단하게 클래스를 사용할 수 있을 것 같아요!

sangxxjin replied:
생성자 생성 제한을 고민하지 않았던 것 같아요. 👍  감사합니다.

ChoiHyunSan commented on backend/src/main/java/com/ll/cafeservice/domain/admin/service/AdminService.java:
```diff
@@ -0,0 +1,40 @@
+import org.springframework.security.crypto.bcrypt.BCryptPasswordEncoder;
+import org.springframework.stereotype.Service;
+
+@Service
+public class AdminService {
+
+    private final AdminRepository adminRepository;
+    private final BCryptPasswordEncoder passwordEncoder;
+
+    @Value("${ADMIN_USERNAME}")
+    private String adminUsername;
+
+    @Value("${ADMIN_PASSWORD}")
+    private String adminPassword;
+
+    public AdminService(AdminRepository adminRepository, BCryptPasswordEncoder passwordEncoder) {
+        this.adminRepository = adminRepository;
+        this.passwordEncoder = passwordEncoder;
+    }
```
간단한 궁금증(?)이 있습니다!
혹시 @RequiredArgsConstructor를 사용 안하시고, 생성자로 받으시는 이유가 있으신가요?

@Value로 받는 값들은 생성자와 무관하게 스프링이 값을 주입해주는 것으로 알고,
final 필드에 대해서는 충분히 에노테이션을 사용해서 받을 수 있는 것으로 알아서요!
(혹시나 제가 잘못된 정보로 이해하고 있다면 수정 부탁드립니다... ㅠ)

문제가 있다고 보는건 아닌데, 컨트롤러에서는 에노테이션을 통해 주입을 받고 있어서 통일성 측면에서 질문드렸습니다. :)

sangxxjin replied:
결론적으로 리팩토링 과정에서 발생한 실수입니다..ㅎ

처음에 @value를 사용하지 않고 직접 값을 넣어서 사용했습니다.
`private final String adminUsername = "~~";`

이렇게 사용하고 있었어서  @requiredargsconstructor를 사용하지 않았었고,
@value를 적용시키면서 같이 적용할 수 있었는데 놓쳤습니다. 감사합니다. 😀

ChoiHyunSan commented on backend/src/test/java/com/ll/cafeservice/domain/admin/service/AdminServiceTest.java:
```diff
@@ -0,0 +1,53 @@
+        admin.encodePassword(encoder);
+    }
+
+    @Test
+    @DisplayName("데이터 삽입 테스트")
+    void testInsertAdminData() {
+        Admin createdAdmin = adminService.createAdmin("testAdmin", "1234");
+        assertEquals(admin.getUsername(), createdAdmin.getUsername());
+        assertTrue(encoder.matches("1234", createdAdmin.getPassword()));
+    }
+
+    @Test
+    @DisplayName("삽입된 데이터 검증 테스트")
+    void testSavedAdminData() {
+        adminService.createAdmin("testAdmin", "1234");
+        Admin foundAdmin = (Admin) adminRepository.findByUsername("testAdmin").orElseThrow();
+        assertEquals(admin.getUsername(), foundAdmin.getUsername());
+        assertTrue(encoder.matches("1234", foundAdmin.getPassword()));
+    }
```
Getter함수를 사용하는 부분이 이 곳만 보여서, 테스트를 위해서 어쩔 수 없이 사용하신 것으로 보여요!

쓰는 것이 좀 더 편하다고 생각이 들기는 하는데 만약 극한으로 안쓰려면 
1. admin의 username은 equals상으로 유니크 값을 가지므로, equals 재정의를 통해 admin과 foundAdmin을 곧바로 비교하게끔 할 수 있을 것 같습니다.

2. 비밀번호는 무조건 encoder가 필요하다 보니, Admin 메서드를 하나 추가해서 사용하는 것을 고려해볼 수 있을 것 같아요. 컨트롤러에서 아직 login 로직이 처리가 안되고 있는 것 같은데 이런 메서드 있으면 테스트에서도 쓰고, 로직 상에서도 쓸 수 있지 않을까요? (사실 불필요하다고 느껴지시면 안쓰셔도 무방한 수준이라고 생각합니다:) )
```java
@Entity
public class Admin {
    private String password;
    
    public boolean validatePassword(PasswordEncoder encoder, String rawPassword) {
        return encoder.matches(rawPassword, this.password);
    }
}
```

sangxxjin replied:
2번으로 적용했습니다. 감사합니다 😊
1. equals는 말 그대로 같은지를 확인하는 메서드라고 생각해서 엄연히 두 데이터는 100% 똑같은 객체다? 라고 보기 어려워서 사용을 안했었습니다.

2. 현재 test코드에서 `encoder.matches(rawPassword, this.password);`랑 같은 작업을 진행하는 부분을 메소드화 시켜주셨는데 아직 관리자 로그인 쪽 진행하지 않았었는데 먼저 반영해도 괜찮을 것 같네요.

ChoiHyunSan reviewed (commented):
작업해주신 전체 부분을 확인했습니다!

논리적으로 오류가 있거나 불필요하게 작성된 것은 없다고 느껴져서 깔끔하게 잘 짜셨다는 감상을 받았습니다:)

감사합니다.

quiemew0 commented:
코드에 대한리뷰라기보다는 협업? 깃헙? 이런것들도 배우고 갑니다🙌🙌
commit 자주 자주하는 습관들여야하는데 아직도 안 익숙하네요 상진님 commit목록들 보고 스스로 반성했어요😂

sangxxjin committed:
refactor: 생성자 생성 제한
본문:
protected 접근 수준으로 변경하여 생성자 생성을 제한했습니다.

sangxxjin committed:
refactor: @RequiredArgsConstructor 적용

sangxxjin committed:
feat: 비밀번호 검증 구현

sangxxjin committed:
test: 비밀번호 검증 적용 및 service 테스트에 맞게 수정
본문:
이전에 setup으로 builder를 사용해서 데이터를 삽입하고 service를 이용해서 데이터를 삽입 했었는데 service 테스트에 맞게 service를 이용해서 데이터를 삽입하고 검증

sangxxjin commented:
@ChoiHyunSan 리뷰 남겨주신 부분 답변 달면서 의견 나누었고 적용했는데 확인 한번 부탁드립니다.

ChoiHyunSan commented:
> @ChoiHyunSan 리뷰 남겨주신 부분 답변 달면서 의견 나누었고 적용했는데 확인 한번 부탁드립니다.

 @sangxxjin 넵 변경사항 확인했습니다. 더 특이사항 없으시면 Merge 진행해주세요!
